samplers/aws/xray/internal/rule.go

```diff
@@ -0,0 +1,213 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal // import "go.opentelemetry.io/contrib/samplers/aws/xray/internal"
+
+import (
+	"sync/atomic"
+	"time"
+
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/trace"
+)
+
+// Rule represents a sampling rule which contains rule properties and reservoir which keeps tracks of sampling statistics of a rule
+type Rule struct {
+	samplingStatistics samplingStatistics
+
+	// reservoir has equivalent fields to store what we receive from service API getSamplingTargets
+	// https://docs.aws.amazon.com/xray/latest/api/API_GetSamplingTargets.html
+	reservoir reservoir
+
+	// equivalent to what we receive from service API getSamplingRules
+	// https://docs.aws.amazon.com/cli/latest/reference/xray/get-sampling-rules.html
+	ruleProperties ruleProperties
+}
+
+type samplingStatistics struct {
+	// number of requests matched against specific rule
+	matchedRequests int64
+
+	// number of requests sampled using specific rule
+	sampledRequests int64
+
+	// number of requests borrowed using specific rule
+	borrowedRequests int64
+}
+
+// stale checks if targets (sampling stats) for a given rule is expired or not
+func (r *Rule) stale(now time.Time) bool {
+	reservoirRefreshTime := r.reservoir.refreshedAt.Add(time.Duration(r.reservoir.interval) * time.Second)
+	return r.samplingStatistics.matchedRequests != 0 && now.After(reservoirRefreshTime)
+}
+
+// snapshot takes a snapshot of the sampling statistics counters, returning
+// samplingStatisticsDocument. It also resets statistics counters.
+func (r *Rule) snapshot(now time.Time) *samplingStatisticsDocument {
+	name := r.ruleProperties.RuleName
+	requests, sampled, borrowed := r.samplingStatistics.matchedRequests, r.samplingStatistics.sampledRequests, r.samplingStatistics.borrowedRequests
```

Comment:
I remember commenting previously that we want to use `atomic.swap` for these get/resets